### DISCUSSION TITLE
Should eglot support 'additionalTextEdits' without 'textEdit'?

### DISCUSSION THREAD

**Discussion by ktfleming** [joaotavora/eglot#981] [Q&A] [answered]
Starting this as a discussion because I'm not yet sure if this is an issue in Eglot or the server in question (typescript-language-server).

From what I can tell of the [LSP spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/), a CompletionItem returned by the server could have an optional `textEdit` as well as any number of `additionalTextEdits`. I don't see anything that specifies that `additionalTextEdits` should only be sent/used if `textEdit` itself is sent. However, the name and documentation for `additionalTextEdits` could be interpreted this way, as the latter mentions a "main edit" (does it make sense for there to be additional edits without a "main edit")?

Eglot currently only applies `additionalTextEdits` if there is also a `textEdit`:

https://github.com/joaotavora/eglot/blob/e835996e16610d0ded6d862214b3b452b8803ea8/eglot.el#L2684

However, typescript-language-server will send a message with `additionalTextEdits`, but no `textEdit`, when selecting an item for completion that has not yet been imported:

```
(:jsonrpc "2.0" :id 285 :result
          (:label "DateTime" :kind 7 :sortText "￿16" :commitCharacters
                  ["." "," "("]
                  :data
                  (:file "/Users/kevin/demo/src/index.ts" :line 109 :offset 20 :entryNames
                         [(:name "DateTime" :source "luxon" :data
                                 (:exportName "DateTime" :moduleSpecifier "luxon" :fileName "/Users/kevin/demo/node_modules/@types/luxon/index.d.ts" :isPackageJsonImport t))])
                  :detail "Auto import from 'luxon'\nclass DateTime" :documentation
                  (:kind "markdown" :value "(long documentation, trimmed for readability)")
                  :additionalTextEdits
                  [(:range
                    (:start
                     (:line 6 :character 0)
                     :end
                     (:line 6 :character 0))
                    :newText "import { DateTime } from \"luxon\";\n")]))
```

This means that selecting the completion will not result in the desired import statement being added. I'm curious if this should be addressed on the Eglot side (support `additionalTextEdits` without `textEdit`) or the server side. Perhaps someone more familiar with the LSP spec would know what the expected behavior is here?

**Comment by joaotavora** [comment, marked as answer]
It'd odd that the server does that, but then again, the LSP standard is pretty odd here, too.  And I think your suggestion that Eglot should support them makes sense.  Can you try this patch?
```diff
diff --git a/eglot.el b/eglot.el
index 4e28de1..887e52c 100644
--- a/eglot.el
+++ b/eglot.el
@@ -2694,16 +2694,16 @@ for which LSP on-type-formatting should be requested."
                                        (eglot--range-region range)))
                             (delete-region beg end)
                             (goto-char beg)
-                            (funcall (or snippet-fn #'insert) newText)))
-                        (when (cl-plusp (length additionalTextEdits))
-                          (eglot--apply-text-edits additionalTextEdits)))
+                            (funcall (or snippet-fn #'insert) newText))))
                        (snippet-fn
                         ;; A snippet should be inserted, but using plain
                         ;; `insertText'.  This requires us to delete the
                         ;; whole completion, since `insertText' is the full
                         ;; completion's text.
                         (delete-region (- (point) (length proxy)) (point))
-                        (funcall snippet-fn (or insertText label)))))
+                        (funcall snippet-fn (or insertText label))))
+                 (when (cl-plusp (length additionalTextEdits))
+                   (eglot--apply-text-edits additionalTextEdits)))
                (eglot--signal-textDocument/didChange)
                (eldoc)))))))))
 
```

**Reply by ktfleming** [reply]
That patch seems to address the issue for me -- now the import lines are added when I select the completion item. 👍

**Reply by joaotavora** [reply]
I've opened a PR, let's see if other people have comments.  https://github.com/joaotavora/eglot/pull/982
